scrapy_zyte_api/providers.py

```diff
@@ -40,33 +46,42 @@ class ZyteApiProvider(PageObjectInputProvider):
         Article,
         ArticleList,
         ArticleNavigation,
+        AnyResponse,
         JobPosting,
     }
 
-    def __init__(self, injector):
-        super().__init__(injector)
-        self._cached_instances: WeakKeyDictionary[Request, Dict] = WeakKeyDictionary()
-
     def is_provided(self, type_: Callable) -> bool:
         return super().is_provided(strip_annotated(type_))
 
     def update_cache(self, request: Request, mapping: Dict[Any, Any]) -> None:
-        if request not in self._cached_instances:
-            self._cached_instances[request] = {}
-        self._cached_instances[request].update(mapping)
+        if request not in self.injector.weak_cache:
+            self.injector.weak_cache[request] = {}
+        self.injector.weak_cache[request].update(mapping)
 
     async def __call__(  # noqa: C901
         self, to_provide: Set[Callable], request: Request, crawler: Crawler
     ) -> Sequence[Any]:
         """Makes a Zyte API request to provide BrowserResponse and/or item dependencies."""
-        # TODO what if ``response`` is already from Zyte API and contains something we need
         results: List[Any] = []
 
+        http_response = None
         for cls in list(to_provide):
-            item = self._cached_instances.get(request, {}).get(cls)
+            item = self.injector.weak_cache.get(request, {}).get(cls)
             if item:
                 results.append(item)
                 to_provide.remove(cls)
+
+            # BrowserResponse takes precedence than HttpResponse
+            elif cls == AnyResponse and BrowserResponse not in to_provide:
+                http_response = self.injector.weak_cache.get(request, {}).get(
+                    HttpResponse
+                )
```

Comment:
How do we ensure that the provider which creates HttpResponse is executed before this provider?

Reply:
By default, `scrapy_poet.page_input_providers.HttpResponseProvider` is given a priority of **500** ([code ref](https://github.com/scrapinghub/scrapy-poet/blob/master/scrapy_poet/downloadermiddlewares.py#L32)), while `scrapy_zyte_api.providers.ZyteApiProvider` only has the recommendation [from the docs](https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/docs/first-steps/scrapy-poet.rst) to be assigned with **1100**.

Hard to ensure as `ZyteApiProvider` isn't assigned with a priority value by default and needs manual user assignment. But if the priority configurations are set right, this case is captured in the [test_provider_any_response_browser_http_response](https://github.com/scrapy-plugins/scrapy-zyte-api/pull/161/files/932262288789a774476ca3d6474caf5d925703dc#diff-864023bf5f9f833678b86d78cf8354385edaf4439d52876ab6b819bfe58d8c35R748-R781) test case.

Reply:
Aha, right. For me it was not clear that the instances are built in the order of provider priority, but indeed that's the case.